lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,351 @@
+
+    private val _endpointStatusFlow = MutableStateFlow<Map<String, EndpointInfo>>(emptyMap())
+    val endpointStatusFlow = _endpointStatusFlow.asStateFlow()
+
+
+    init {
+        observeEndpointStatusFlow()
+    }
+
+    fun start() {
+        startAdvertising()
+        startDiscovery()
+    }
+
+    fun stop() {
+        connectionsClient.stopAdvertising()
+        connectionsClient.stopDiscovery()
+        connectionsClient.stopAllEndpoints()
+        _endpointStatusFlow.value = emptyMap()
+        log("Stopped advertising, discovery, all connections, and cleared endpoints")
+    }
+
+
+    private fun startAdvertising() {
+        val advertisingOptions = AdvertisingOptions.Builder().setStrategy(strategy).build()
+        connectionsClient.startAdvertising(
+            name, serviceId, connectionLifecycleCallback, advertisingOptions
+        ).addOnSuccessListener {
+            log("Started advertising successfully")
+        }.addOnFailureListener { e ->
+            log("Failed to start advertising", e)
+        }
+    }
+
+    private fun startDiscovery() {
+        val discoveryOptions = DiscoveryOptions.Builder().setStrategy(strategy).build()
+        connectionsClient.startDiscovery(
+            serviceId, endpointDiscoveryCallback, discoveryOptions
+        ).addOnSuccessListener {
+            log("Started discovery successfully")
+        }.addOnFailureListener { e ->
+            log("Failed to start discovery", e)
+        }
+    }
+
+    private val connectionLifecycleCallback = object : ConnectionLifecycleCallback() {
+        override fun onConnectionInitiated(endpointId: String, connectionInfo: ConnectionInfo) {
+            log("Connection initiated with endpoint: $endpointId")
+            connectionsClient.acceptConnection(endpointId, payloadCallback)
+            updateEndpointStatus(endpointId, EndpointStatus.CONNECTING)
+        }
+
+        override fun onConnectionResult(endpointId: String, result: ConnectionResolution) {
+            if (result.status.isSuccess) {
+                log("Successfully connected to endpoint: $endpointId")
+                updateEndpointStatus(endpointId, EndpointStatus.CONNECTED)
+                sendMmcpPingPacket(endpointId)
+            } else {
+                log("Failed to connect to endpoint: $endpointId. Reason: ${result.status}")
+                updateEndpointStatus(endpointId, EndpointStatus.DISCONNECTED)
+            }
+        }
+
+        override fun onDisconnected(endpointId: String) {
+            log("Disconnected from endpoint: $endpointId")
+            updateEndpointStatus(endpointId, EndpointStatus.DISCONNECTED)
+        }
+    }
+
+    private val endpointDiscoveryCallback = object : EndpointDiscoveryCallback() {
+        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
+            log("Endpoint found: $endpointId")
+            addEndpoint(endpointId)
+            checkAndInitiateConnection(endpointId)
```

Comment:
Connections should __only__ be initiated by observing the status. We could discover more nodes than we want to connect to,  this function will _wrongly_ attempt to connect to every single node discovered.